chainer/link.py

```diff
@@ -298,31 +299,56 @@ def register_persistent(self, name):
         self._persistent.add(name)
         self._params.discard(name)
 
-    def copy(self):
+    def copy(self, mode='share'):
         """Copies the link hierarchy to new one.
 
-        The whole hierarchy rooted by this link is copied. The copy is
-        basically shallow, except that the parameter variables are also
-        shallowly copied. It means that the parameter variables of copied one
-        are different from ones of original link, while they share the data and
-        gradient arrays.
+        The whole hierarchy rooted by this link is copied. There are three
+        modes to perform copy. Please see the document for the argument
+        ``mode`` below.
 
         The name of the link is reset on the copy, since the copied instance
         does not belong to the original parent chain (even if exists).
 
+        Args:
+            mode (str): It should be either ``init``, ``copy``, or ``share``.
+                ``init`` means parameter variables under the returned link
+                object is re-initialized by calling their
+                :meth:`~chainer.Parameter.initialize` method, so that all the
+                parameters may have different initial values from the original
+                link.
+                ``copy`` means that the link object is deeply copied, so that
+                its parameters are not re-initialized but are also deeply
+                copied. Thus, all parameters have same initial values but can
+                be changed independently.
+                ``share`` means that the link is shallowy copied, so that its
+                parameters' arrays are shared with the original one. Thus,
+                their values are changed synchronously. The default ``mode``
+                is ``share``.
+
         Returns:
             Link: Copied link object.
 
         """
-        ret = copy.copy(self)
-        ret._params = set(self._params)
-        ret._persistent = set(self._persistent)
-        ret.name = None
-        d = ret.__dict__
-        for name in ret._params:
-            d[name] = copy.copy(d[name])
-            d[name].grad = None
-        return ret
+        if mode not in ['init', 'copy', 'share']:
+            raise ValueError(
+                'The \'mode\' argument should be either \'init\','
+                '\'copy\', or \'share\'. But {} was given.'.format(mode))
+        if mode == 'share':
+            ret = copy.copy(self)
+            ret._params = set(self._params)
+            ret._persistent = set(self._persistent)
+            ret.name = None
+            d = ret.__dict__
+            for name in ret._params:
+                d[name] = copy.copy(d[name])
+                d[name].grad = None
+            return ret
+        elif mode == 'copy' or mode == 'init':
+            ret = copy.deepcopy(self)
```

Comment:
What happens when the mode is `copy` and the link includes uninitialized parameters?

Reply:
The elements of copied and original Sequential become different objects and both keep uninitialized.